インクリメンタルサーチの実装

fujitami opened the pull request:
# WHAT
インクリメンタルサーチによるユーザ検索機能の実装。

# WHY
ユーザ検索時の利便性向上のため。

![default](https://user-images.githubusercontent.com/39646282/42882887-f21c1308-8ad4-11e8-9596-e5843bd33ac0.gif)

fujitami committed:
first commit

fujitami committed:
Update users_controller.rb

ユーザ検索処理
API設定

fujitami committed:
Update groups/_form.html.haml

ユーザ検索用テキストフィールドの設置
検索結果のリスト表示(チェックボックスの削除)

fujitami committed:
Add index.json.jbuilder

fujitami committed:
Add index.js

インクリメンタルサーチによるユーザ検索機能

fujitami committed:
Update views

検索結果の追加ボタンの実装
チャットメンバー欄の削除ボタンの実装

MrUrakou commented on app/assets/javascripts/index.js:
```diff
@@ -0,0 +1,70 @@
+    var removeObj = $('li').eq(`${index}`).remove();
+  });
+
+  $('#chat-group-users').on('click', '.chat-group-user__btn--remove', function(e) {
```
ここは.chat-group-user__btn--removeではなく.js-remove-btnで指定するのが正しいです。
何故ならば見た目を調整するclassとjsで使用するclassは分けたほうが後々のメンテナンスがしやすいからです（なのでわざわざ複数のclassを付与しています。）addの方も同様に修正お願いします。

MrUrakou commented on app/controllers/users_controller.rb:
```diff
@@ -1,6 +1,11 @@
 class UsersController < ApplicationController
 
   def index
+    @users = User.where('name LIKE(?)', "%#{params[:keyword]}%").limit(20)
```
limit(20)にしている根拠はありますか？
もしそうするなら定数に代入して使い回したいです。
そうしないとマジックナンバー化のリスクがあるからです。
あとcurrent_userはデフォルトでgroupに入る仕様なので検索されないようにしましょう。

fujitami replied:
深く考えておりませんでした……。
一旦削除し、のちに表示件数を絞る適切な理由が生じたら、それを明確にした定数を作成して設定いたします。

MrUrakou commented on app/views/groups/_form.html.haml:
```diff
@@ -11,31 +11,18 @@
-      <div id='user-search-result'></div>
-      </div>
+    .chat-group-form__field--left
+      %label.chat-group-form__label{for: "chat_group_チャットメンバーを追加"} チャットメンバーを追加
```
以下、f.labelなどhelper使えるところはhelper使いましょう。

MrUrakou commented on app/assets/javascripts/index.js:
```diff
@@ -0,0 +1,70 @@
+$(function() {
+var search_list = $('.user-search-index');
```
idに対してピンポイントでappendしたいですね。

fujitami committed:
Update _form.html.haml

f.Formヘルパーの使用

fujitami committed:
Update users_controller.rb

不要なlimitの削除

fujitami committed:
Update index.js

イベント・変数の指定先要素の調整

TOMARUMARU reviewed (commented):
検索はcurrent_userを除いた検索結果になっていますか？

fujitami committed:
Update users_controller.rb

検索結果からcurrent_userを除外

MrUrakou commented on app/assets/javascripts/index.js:
```diff
@@ -0,0 +1,72 @@
+$(function() {
+var search_list = $('#user-search-result');
+var chat_users = $('#chat-group-users');
```
この2行、インデント深くしましょう。

MrUrakou commented on app/assets/javascripts/index.js:
```diff
@@ -0,0 +1,72 @@
+  $('#chat-group-users').on('click', '.js-remove-btn', function(e) {
+    e.stopPropagation();
+    var index = $('.js-remove-btn').index(this);
+    $('div.js-chat-member').eq(`${index}`).remove();
```
対象のDOM要素の親要素をremoveするという記述があれば、ワンライナーに出来ると思います。

MrUrakou commented on app/assets/javascripts/index.js:
```diff
@@ -0,0 +1,72 @@
+    addChatUser(id, name);
+
+    var index = $('.user-search-add').index(this);
+    var removeObj = $('li').eq(`${index}`).remove();
```
この変数って使ってますか？

fujitami replied:
すみません、不要になった際に消し忘れました。

MrUrakou reviewed (commented):
完成品を参考に、デフォルトでuser(あるいはcurrent_user)を表示させてください。

fujitami committed:
Update _form.html.haml

デフォルトでcurrent_userをチャットメンバー欄に表示

fujitami committed:
Update index.js

変数のインデント調整
appendNoUserの修正(ul要素追加)
ユーザ検索結果及びチャットメンバー欄において、選択したユーザを一覧から削除する機能の記述を修正
